### DISCUSSION TITLE
Import statement style

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#2914] [Q&A]
In the past we decided on the following import statement style:
1. Use dependent crates directly e.g., 
  - DO: `use hashes::Hash` 
  - DON'T:  `use crate::hashes::Hash`
2. Import crate types using their exact location not the re-exports e.g., 
  - DO: `use crate::blockdata::transaction::Transaction`
  - DON'T: `use crate::Transaction`
 
I can't remember why we did (1), we did (2) with the aim of reducing the size of diffs when imports changed.

I would like to propose changing the style to use import statements that mirror the way downstream users should use the crate eg, import everything using the most terse re-export 

`use crate::{amount, Transaction, Amount} // amount module for error types `

This will mean a lot of code churn but buys us two big things:

1. Make the transition to `primitives` easier because all the import statements will then work without needing to be changed
2. Allow use to proof the re-exports (and have some confidence that downstream users are doing the same) so we can eventually remove all the code organisation things (eg `blockdata::transaction::Transaction`). 

## Justification

- We have an unwritten aim (in my mind at least) of decoupling the public API from the code organisation. 
- I have a gut feeling that users of `rust-bitcoin` copy the import statements that we use so there are downstream users using for example `blockdata`.
- If we implement this proposal it will be a massive, ugly diff that is 100% internal and can be reasonably soundly reviewed and merged. Then the `primitives` PR, which is going to be very difficult to review, will become considerably smaller.

**Comment by Kixunil** [comment]
1. I believe it's just to avoid redundancy
2. looks like a weird rule, are you sure we have that rule? Could you be mistaking it with preferrring `super::foo` whenever `super` != `crate` || `foo` is *defined* in `crate`?

I want to stress that the `super` rule is very helpful in code moves to other crates as well. Also I think `primitives` should have very similar structure as `bitcoin` but maybe `primitives` ~ `blockdata`.

**Reply by tcharding** [reply]
> 2. looks like a weird rule, are you sure we have that rule? 

You mean rule (2)? I went to quickly check using `Transaction` and found that we are non-uniform on use of `blockdata::` but _almost_ uniform in not using `Transaction` directly from the crate root.

```bash
gg 'use crate' | grep Transaction | grep -v blockdata
bitcoin/src/bip152.rs:18:use crate::{block, Block, BlockHash, Transaction};
bitcoin/src/consensus/encode.rs:24:use crate::bip152::{PrefilledTransaction, ShortId};
bitcoin/src/consensus/encode.rs:36:use crate::transaction::{Transaction, TxIn, TxOut};
bitcoin/src/consensus/validation.rs:16:use crate::transaction::{OutPoint, Transaction, TxOut};
bitcoin/src/crypto/sighash.rs:24:use crate::{transaction, Amount, Script, ScriptBuf, Sequence, Transaction, TxIn, TxOut};
bitcoin/src/merkle_tree/block.rs:21:use crate::transaction::{Transaction, Txid};
bitcoin/src/p2p/message.rs:546:    use crate::bip152::BlockTransactionsRequest;
bitcoin/src/p2p/message.rs:559:    use crate::transaction::Transaction;
bitcoin/src/psbt/error.rs:11:use crate::transaction::Transaction;
bitcoin/src/psbt/map/global.rs:11:use crate::transaction::Transaction;
bitcoin/src/psbt/map/input.rs:22:use crate::transaction::{Transaction, TxOut};
bitcoin/src/psbt/mod.rs:29:use crate::transaction::{self, Transaction, TxOut};
bitcoin/src/psbt/serialize.rs:22:use crate::transaction::{Transaction, TxOut};
```

**Reply by Kixunil** [reply]
Yes, interesting.

**Comment by apoelstra** [comment]
For my part I don't feel strongly about any of this. *In my own code* I try to use a single module prefix like `bitcoin::Transaction` because it helps clue me in to where a type was defined (at least, whether it was defined in the same module that I'm reading) and sometimes it reads nicer.

I designed the existing hashes API with this in mind, so you could write `sha256::Hash` etc. I believe that API was a mistake, but I still like my coding style.

I don't mind having rules here, but AFAIK we have not had any rules in the past beyond my personal preference (and my having written much of the original codebase). Other than avoiding wildcards I don't care what the rules are. Feel free to remind me if I run afoul of them.

**Reply by tcharding** [reply]
Cool, I'm not really remembering when or how we got to the set of rules that is in my head, I do know that I'm anal as f*** and have tried to make things uniform since day one in the project, much to the annoyance of many - one day I may calm down, I don't see it any time soon though, sorry not sorry.

I'm good to introduce and start enforcing use of `super::foo` anytime that `super != crate`. I'm going to go patch `bdk` and maybe others to at least remove `blockdata::` so we can consider not re-exporting from that path sometime soon-ish.

**Reply by tcharding** [reply]
I did `bdk`, `rust-lightning`, and `fedimint`.

**Comment by Kixunil** [comment]
Speaking about rules, I'd love them to be compatible with my merge tool. :) (haven't checked whether these actually are)
